### ISSUE TITLE
oce.plot.ts not working with xlim provided

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1508]
I'm having a hard time producing a MWE, but the data that I am seeing this behaviour on is one that @dankelley is currently looking at anyway. I was trying to look at a zoomed-in region of the sea021m49 glider data, but `oce.plot.ts()` keeps saying that there is no data in the range (even though I know there is). I suspect the problem is with the `trim_ts()` function, which is in C, so I'm pointing @dankelley at the issue to see if he can reproduce.

```r
library(oceanglider)
file <- "~/Dropbox/sea021m49/netcdf/GLI2018_SEA021_049DM_L1.nc"
g <- read.glider(file)
head(g[["time"]], 10)
## The above shows that we have a few (4) points from 2018-07-12 at the start,
## followed by data on 2019-02-22, so we subset for just the later data.
start <- as.POSIXct("2019-02-21")
g <- subset(g, time > start)

focus <- as.POSIXct(c('2019-02-27', '2019-03-05'), tz='UTC')
oce.plot.ts(g[['time']], g[['pressure']], xlim=focus)
```
Note the appearance of a `NULL` before the warnings -- I can't figure out where that is coming from and why it is being printed to the console:
```
NULL
Warning messages:
1: In min.default(NA_real_, na.rm = TRUE) :
  no non-missing arguments to min; returning Inf
2: In max.default(NA_real_, na.rm = TRUE) :
  no non-missing arguments to max; returning -Inf
3: In min(x) : no non-missing arguments to min; returning Inf
4: In max(x) : no non-missing arguments to max; returning -Inf
5: In oce.plot.ts(g[["time"]], g[["pressure"]], xlim = focus) :
  no valid data for 'g[["pressure"]]'
```

**Comment by richardsc** [comment]
Below is the reason I am trying to do this -- there is something weird going on in the glider data where there are some "downcast" points being stuck into the data stream that I think are bogus:

![image](https://user-images.githubusercontent.com/233584/54477872-f5025180-47ea-11e9-8ee5-b8e20171a6af.png)

**Comment by dankelley** [comment]
I cannot reproduce this. I created a test file in the `oce-issues` repo, using the Dropbox as indicated (that file has md5 9ed659827634c18f2233918adbcc4335).  I don't get any warnings.

Can you maybe copy that file to your own dropbox and try the code on the repo?

I can't remember how to make R spit out the local settings, but maybe we have different values of some locale or something. But I suspect you're looking at a different file. (???)

**Comment by richardsc** [comment]
Sorry -- I didn't copy the full code. I was also doing a subset on salinity (to get rid the crazy high ones). I'll update the oce-issues test to include that (which *does* fail, for reasons I can't discern):

```r
library(oceanglider)
if (!exists("g")) {
    file <- "/data/archive/glider/2019/sx/sea021m49/netcdf/GLI2018_SEA021_049DM_L1.nc"
    g <- read.glider(file)
    head(g[["time"]], 10)
    ## The above shows that we have a few (4) points from 2018-07-12 at the start,
    ## followed by data on 2019-02-22, so we subset for just the later data.
    start <- as.POSIXct("2019-02-21")
    g <- subset(g, time > start)
    g  <- subset(g, salinity < 40)
}

par(mfrow=c(2,1))
focus <- as.POSIXct(c('2019-02-27', '2019-03-05'), tz='UTC')
oce.plot.ts(g[['time']], g[['pressure']], xlim=focus)
focus <- as.POSIXct(c('2019-02-27', '2019-02-28'), tz='UTC')
oce.plot.ts(g[['time']], g[['pressure']], xlim=focus)
```

**Comment by richardsc** [comment]
So now I'm really confused ... maybe the problem is actually in the `subset()` call? So that it is messing up the object somehow?

But as you can see above, if I don't use `oce.plot.ts()` but just `plot()` it works fine ...

**Comment by dankelley** [comment]
This ought to work now, in "develop" branch commit 21a163fffd5f3ce199287e5761997f00f87e26ea

The problem was (I think) that the `subset,glider-method` was introducing `NA` values in the times. That messed up the `trim_ts` C++ code. But, heck, the only purpose of that code was to try to save some memory on RAM-limited machines (by not creating a long boolean vector, but rather a shorter indexing vector). But, as can be seen by inspection of that C++ code (`src/trim.cpp`), it would be fairly tricky to fix it. Near the start (line 34), I create a fuzzy-inequality quantity `epsilon` based on the first two `x` values. Well, to fix this, we would have to find the first two adjacent non-`NA` values. And I suppose other weirdnesses might arise also.

The code I added seems to work fine. Evidently, R is very clever with inequalities, and it skips over `NA` values in boolean lists used for subset operations. Good old R.

I do not think it is really worth it to fix `trim_ts` because it is not a function that is intended for the user. It is not documented, and it's not in the NAMESPACE.  I'll leave the `src/trim.cpp` code in existence for a while, but I plan to remove it and the various places where it is hooked up to the R code, once this issue gets closed.

PS. as for why `glider,subset-method` is giving `NA` values, that's something I'll look into. I'll make up a small test for it.  (I'll maybe start by writing `as.glider()`, if that doesn't exist yet, because then I can make up fake data inside the test suite. I could use a snippet of real data, if I new of some data that were open enough that I would feel OK in using them.)

**Comment by richardsc** [comment]
Correct me if I'm wrong, but the fix for this was just to do the trimming with base R instead of calling out to a C function?

I'd say this is fixed then, though I don't know if there are any other consequences to the change (looks to me like the worst that can happen is the plotting might be slowed down a bit for very large datasets).

**Comment by richardsc** [comment]
One comment: The old `trim_ts` had an argument to allow the trim to be slightly longer than the actual `xlim`. This meant that the trimmed time series would still "go off the edge" of the plot.

The new behaviour trims to the xlim, but because the axes are automatically extended (by 4% or whatever it is), it looks as though the series doesn't continue past the xlims.

**Comment by richardsc** [comment]
Fixed in 54163db.
